File: benchmarks/comparison/cpp/14_hyperbolic_pde.cpp
// Hyperbolic PDE Solver Benchmark
// Same algorithm as Seen: 1D catenary/cable equation
// grid=10000, steps=500, dx=4.0/grid, dt=0.0001
#include <cstdio>
#include <cstdint>
#include <cmath>
#include <vector>
#include <chrono>

static double run_hyperbolic_pde(int grid_size, int timesteps) {
    double dx = 4.0 / (double)grid_size;
    double dt = 0.0001;

    std::vector<double> u((size_t)grid_size, 0.0);
    std::vector<double> v((size_t)grid_size, 0.0);
    std::vector<double> u_new((size_t)grid_size, 0.0);

    // Set initial condition: catenary centered in domain
    for (int i = 0; i < grid_size; i++) {
        double x = -2.0 + (double)i * dx;
        u[(size_t)i] = std::cosh(x) - 1.0;
        v[(size_t)i] = 0.0;
    }

    // Time-stepping loop
    for (int t = 0; t < timesteps; t++) {
        double time = (double)t * dt;

        // Interior points update
        for (int j = 1; j < grid_size - 1; j++) {
            double x = -2.0 + (double)j * dx;

            // Catenary restoring force
            double catenary_force = std::cosh(x) - u[(size_t)j];

            // Tension term using sinh
            double tension = 1.0;
            if (std::fabs(x) > 0.0001) {
                tension = std::sinh(x) / x;
            }

            // Damping using tanh (saturates at +-1)
            double damping = -0.1 * std::tanh(v[(size_t)j]);

            // Thermal decay using exp
            double thermal = std::exp(-0.01 * time);

            // Laplacian (second derivative)
            double laplacian = (u[(size_t)(j + 1)] - 2.0 * u[(size_t)j] + u[(size_t)(j - 1)]) / (dx * dx);

            // Update velocity and position
            double accel = tension * laplacian + catenary_force * thermal + damping;
            v[(size_t)j] = v[(size_t)j] + accel * dt;
            u_new[(size_t)j] = u[(size_t)j] + v[(size_t)j] * dt;
        }

        // Boundary conditions: fixed ends
        u_new[0] = std::cosh(-2.0) - 1.0;
        u_new[(size_t)(grid_size - 1)] = std::cosh(2.0) - 1.0;

        // Copy u_new -> u
        for (int k = 0; k < grid_size; k++) {
            u[(size_t)k] = u_new[(size_t)k];
        }
    }

    // Compute checksum: sum of final grid values plus some log evaluations
    double checksum = 0.0;
    for (int m = 0; m < grid_size; m++) {
        checksum += u[(size_t)m];
        double abs_val = std::fabs(u[(size_t)m]) + 1.0;
        checksum += std::log(abs_val);
    }

    return checksum;
}

int main() {
    int grid_size = 10000;
    int timesteps = 500;

    printf("Hyperbolic PDE Solver Benchmark\n");
    printf("Grid size: %d, Timesteps: %d\n", grid_size, timesteps);

    int warmup_runs = 3;
    printf("Warming up (%d runs at 100 timesteps)...\n", warmup_runs);
    for (int w = 0; w < warmup_runs; w++) {
        (void)run_hyperbolic_pde(grid_size, 100);
    }

    printf("Running measured iterations...\n");
    int iterations = 5;
    double min_time = 1e18;
    double result = 0.0;

    for (int iter = 0; iter < iterations; iter++) {
        auto start = std::chrono::high_resolution_clock::now();
        double checksum = run_hyperbolic_pde(grid_size, timesteps);
        auto end = std::chrono::high_resolution_clock::now();
        double elapsed = std::chrono::duration<double, std::milli>(end - start).count();
        if (elapsed < min_time) {
            min_time = elapsed;
            result = checksum;
        }
    }

    printf("Checksum: %.6f\n", result);
    printf("Min time: %.6f ms\n", min_time);
    return 0;
}

File: benchmarks/comparison/zig/14_hyperbolic_pde.zig
// Hyperbolic PDE Solver Benchmark
// 1D catenary/cable equation with hyperbolic function evaluation
// grid=10000, steps=500, dx=4.0/grid, dt=0.0001
const std = @import("std");
const math = std.math;

fn runHyperbolicPde(allocator: std.mem.Allocator, grid_size: usize, timesteps: usize) !f64 {
    const grid_f: f64 = @floatFromInt(grid_size);
    const dx: f64 = 4.0 / grid_f;
    const dt: f64 = 0.0001;

    const u = try allocator.alloc(f64, grid_size);
    defer allocator.free(u);
    const v = try allocator.alloc(f64, grid_size);
    defer allocator.free(v);
    const u_new = try allocator.alloc(f64, grid_size);
    defer allocator.free(u_new);

    // Set initial condition: catenary centered in domain
    for (0..grid_size) |i| {
        const i_f: f64 = @floatFromInt(i);
        const x = -2.0 + i_f * dx;
        u[i] = math.cosh(x) - 1.0;
        v[i] = 0.0;
    }

    // Time-stepping loop
    for (0..timesteps) |t| {
        const t_f: f64 = @floatFromInt(t);
        const time = t_f * dt;

        // Interior points update
        var j: usize = 1;
        while (j < grid_size - 1) : (j += 1) {
            const j_f: f64 = @floatFromInt(j);
            const x = -2.0 + j_f * dx;

            // Catenary restoring force: cosh(x)
            const catenary_force = math.cosh(x) - u[j];

            // Tension term using sinh: sinh(x)/x gives tension profile
            var tension: f64 = 1.0;
            if (@abs(x) > 0.0001) {
                tension = math.sinh(x) / x;
            }

            // Damping using tanh (saturates at +-1)
            const damping = -0.1 * math.tanh(v[j]);

            // Thermal decay using exp
            const thermal = @exp(-0.01 * time);

            // Laplacian (second derivative)
            const laplacian = (u[j + 1] - 2.0 * u[j] + u[j - 1]) / (dx * dx);

            // Update velocity and position
            const accel = tension * laplacian + catenary_force * thermal + damping;
            v[j] = v[j] + accel * dt;
            u_new[j] = u[j] + v[j] * dt;
        }

        // Boundary conditions: fixed ends
        u_new[0] = math.cosh(@as(f64, -2.0)) - 1.0;
        u_new[grid_size - 1] = math.cosh(@as(f64, 2.0)) - 1.0;

        // Copy u_new -> u
        @memcpy(u, u_new);
    }

    // Compute checksum: sum of final grid values plus some ln evaluations
    var checksum: f64 = 0.0;
    for (0..grid_size) |m| {
        checksum += u[m];
        // Also exercise ln on positive values
        const abs_val = @abs(u[m]) + 1.0;
        checksum += @log(abs_val);
    }

    return checksum;
}

pub fn main() !void {
    const stdout = std.fs.File.stdout().deprecatedWriter();
    const allocator = std.heap.page_allocator;

    const grid_size: usize = 10000;
    const timesteps: usize = 500;

    try stdout.print("Hyperbolic PDE Solver Benchmark\n", .{});
    try stdout.print("Grid size: {d}, Timesteps: {d}\n", .{ grid_size, timesteps });

    // Warmup
    const warmup_runs = 3;
    try stdout.print("Warming up ({d} runs at 100 timesteps)...\n", .{warmup_runs});
    for (0..warmup_runs) |_| {
        _ = try runHyperbolicPde(allocator, grid_size, 100);
    }

    // Measured iterations
    try stdout.print("Running measured iterations...\n", .{});
    const iterations = 5;
    var min_time: f64 = 1e18;
    var result: f64 = 0.0;

    for (0..iterations) |_| {
        var timer = try std.time.Timer.start();
        const checksum = try runHyperbolicPde(allocator, grid_size, timesteps);
        const elapsed_ns = timer.read();
        const elapsed_ms = @as(f64, @floatFromInt(elapsed_ns)) / 1_000_000.0;
        if (elapsed_ms < min_time) {
            min_time = elapsed_ms;
            result = checksum;
        }
    }

    const grid_f: f64 = @floatFromInt(grid_size);
    const steps_f: f64 = @floatFromInt(timesteps);
    const grid_points_per_step = grid_f * steps_f;

    try stdout.print("Checksum: {d:.6}\n", .{result});
    try stdout.print("Min time: {d:.6} ms\n", .{min_time});
    try stdout.print("Grid-steps per second: {d:.6} million\n", .{grid_points_per_step / (min_time / 1000.0) / 1000000.0});
}
